### PR TITLE
Enable reading validation for derived type constraint annotation

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

*This pull request fixes enable the resource reading derived type constraint annotation validation.*

### Description

*Enable the resource reading derived type constraint annotation validation.*

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs [hunk 1]
```diff
@@ -871,6 +871,15 @@ in propertyAndAnnotationCollector.GetODataPropertyAnnotations(propertyName))
                 propertyTypeName = (string)propertyAnnotation.Value;
             }
 
+            if (propertyTypeName != null)
+            {
+                var validater = propertyAndAnnotationCollector.GetDerivedTypeValidator(propertyName);
```
Comment (KanishManuja-MS) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs]: >validater [](start = 20, length = 9)

It should be validator.

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs [hunk 2]
```diff
@@ -1889,6 +1899,12 @@ private void StartReadingResource()
             // Resolve the type name
             this.ApplyResourceTypeNameFromPayload(currentResource.TypeName);
 
+            // Validate type with derived type validator if available
+            if (this.CurrentDerivedTypeValidator != null && !(this.ReadingDelta && this.CurrentResourceDepth == 0))
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs]: >this.ReadingDelta [](start = 62, length = 17)

why don't we validate derived type for delta?

File: src/Microsoft.OData.Core/ODataReaderCore.cs [hunk 3]
```diff
@@ -338,6 +338,18 @@ protected ResourceSetWithoutExpectedTypeValidator CurrentResourceSetValidator
...
 
+        /// <summary>
+        /// Validator to validate the derived type constraint.
+        /// </summary>
+        protected DerivedTypeValidator CurrentDerivedTypeValidator
+        {
+            get
+            {
+                Debug.Assert(this.State == ODataReaderState.ResourceStart, "CurrentDerivedTypeValidator should only be called while reading a resource.");
+                return this.ParentScope == null ? null : this.ParentScope.DerivedTypeValidator;
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/ODataReaderCore.cs]: 1.  We do the resource type validation when we are reading the resource itself.
2. The "@odata.type" is read at the ResourceStart, we should validate it as early as possible.
3. The derived type annotation is from either top level navigation source, or the nested resource (nested property). For all the resource at a certain navigation source, we use one validator.

File: src/Microsoft.OData.Core/ODataReaderCore.cs [hunk 4]
```diff
@@ -513,6 +525,11 @@ protected void EnterScope(Scope scope)
                 scope.ResourceTypeValidator = new ResourceSetWithoutExpectedTypeValidator(scope.ResourceType);
             }
 
+            if (scope.State == ODataReaderState.ResourceSetStart)
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Core/ODataReaderCore.cs]: I need more discussion about the delta scenario.

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        private IEnumerable<string> derivedTypeConstraints;
+        private IEdmType expectedType;
+        private string resourceKind;
+        private string resourceName;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="expectedType">The exptected type.</param>
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/DerivedTypeValidator.cs]: >exptected [](start = 43, length = 9)

nit: spelling ("expected")

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        /// <param name="resourceTypeName">The full type name of the resource.</param>
+        internal void ValidateResourceType(string resourceTypeName)
+        {
+            if (resourceTypeName == null)
+            {
+                return;
+            }
+
+            if (this.expectedType != null && this.expectedType.FullTypeName() == resourceTypeName)
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/DerivedTypeValidator.cs]: >expectedType [](start = 21, length = 12)

Do we have cases where expectedType is null?

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            ValidateResourceType(resourceType.FullTypeName());
+        }
+
+        /// <summary>
+        /// Validates the full type name of a resource.
+        /// </summary>
+        /// <param name="resourceTypeName">The full type name of the resource.</param>
+        internal void ValidateResourceType(string resourceTypeName)
+        {
+            if (resourceTypeName == null)
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/DerivedTypeValidator.cs]: >resourceTypeName [](start = 16, length = 16)

Do we have cases where we would call this with a null resourceTypeName?

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs (hunk 5 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/DerivedTypeValidator.cs]: #Resolved.

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs (hunk 7 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/DerivedTypeValidator.cs]: At top level , ODL supports to create reader without the expected resource type. for example, see https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataMessageReader.cs#L316-L319:

        /// <summary>Creates an <see cref="T:Microsoft.OData.ODataReader" /> to read a resource set.</summary>
        /// <returns>The created reader.</returns>
        public ODataReader CreateODataResourceSetReader()
        {
            return this.CreateODataResourceSetReader(/*entitySet*/null, /*expectedResourceType*/null);
        }

**Comment by mikepizzo** [comment]
A couple questions for you inline, but I'm good with this.

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs (hunk 6 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/DerivedTypeValidator.cs]: At top level , ODL supports to create reader without the expected resource type. for example, see https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataMessageReader.cs#L316-L319:

        /// <summary>Creates an <see cref="T:Microsoft.OData.ODataReader" /> to read a resource set.</summary>
        /// <returns>The created reader.</returns>
        public ODataReader CreateODataResourceSetReader()
        {
            return this.CreateODataResourceSetReader(/*entitySet*/null, /*expectedResourceType*/null);
        }

File: src/Microsoft.OData.Core/DerivedTypeValidator.cs (hunk 7 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/DerivedTypeValidator.cs]: :thumbsup: 

---
In reply to: [252894010](https://github.com/OData/odata.net/pull/1376#discussion_r252894010) [](ancestors = 252894010)

**Comment by mikepizzo** [comment]
Looks good to me; thanks for the follow-up!

**Pull request merged**